library/ecp.c

```diff
@@ -767,6 +767,63 @@ int mbedtls_ecp_point_write_binary( const mbedtls_ecp_group *grp,
     return( ret );
 }
 
+#if defined(MBEDTLS_ECP_SHORT_WEIERSTRASS_ENABLED)
+static int mbedtls_ecp_sw_derive_y( const mbedtls_ecp_group *grp,
+                                    const mbedtls_mpi *X,
+                                    mbedtls_mpi *Y,
+                                    int parity_bit )
+{
+    /* y^2 = x^3 + ax + b
+     * sqrt(w) = w^((p+1)/4) mod p   (for prime p where p = 3 mod 4) */
```

Comment:
I think here we should add a comment that this method for extracting square roots doesn't ensure that `w` was indeed a square, so this function will return garbage in Y in case X does not correspond to a point on the curve. Which is OK, as long as it's made explicit.

Reply:
I agree, but I think there's a difference between accepting a garbage value for Y, and generating one. So, I agree the public documentation doesn't need to be changed, but I'd like us to be clear internally about the limitations of the algorithms we're using.

Also, I don't want someone to copy-paste that code and use it in another context thinking this is a reliable method for extracting square roots mod p (as soon as p == 3 mod 4). I mean I'm not even sure it's general knowledge that about half of the values mod p have no square root, and the comment as it stands could certainly mislead someone into thinking that all values mod p have square roots (because it give a formula for computing them). I just want the comment here to be complete and mention all the restrictions of the formula we're using.

Reply:
:rocket:

Reply:
> I agree, but I think there's a difference between accepting a garbage value for Y, and generating one.  [...]
> 
> Also, I don't want someone to copy-paste that code and use it in another context thinking this is a reliable method for extracting square roots mod p (as soon as p == 3 mod 4). [...]
>
> I just want the comment here to be complete and mention all the restrictions of the formula we're using.

Excellent points.